### ISSUE TITLE
with openssl 3.0.1 have problem

### ISSUE THREAD

**Issue by spacexnasa** [Snapchat/KeyDB#392]
make[1]: Entering directory '/root/KeyDB/src'
    CC Makefile.dep
    CC release.o
    CC tls.o
tls.cpp: In function ‘void tlsInit()’:
tls.cpp:158:26: error: too few arguments to function ‘const char* ERR_lib_error_string(long unsigned int)’
     ERR_lib_error_string();
                          ^
In file included from tls.cpp:41:0:
/usr/include/openssl/err.h:445:13: note: declared here
 const char *ERR_lib_error_string(unsigned long e);
             ^~~~~~~~~~~~~~~~~~~~
tls.cpp:159:5: error: ‘SSL_int_init’ was not declared in this scope
     SSL_int_init();
     ^~~~~~~~~~~~
tls.cpp:159:5: note: suggested alternative: ‘SSL_in_init’
     SSL_int_init();
     ^~~~~~~~~~~~
     SSL_in_init
tls.cpp: In function ‘int tlsConfigure(redisTLSContextConfig*)’:
tls.cpp:359:24: error: cannot convert ‘DH* {aka dh_st*}’ to ‘BIGNUM* {aka bignum_st*}’ for argument ‘1’ to ‘void BN_free(BIGNUM*)’
              BN_free(dh);
                        ^
In file included from /usr/include/openssl/crypto.h:36:0,
                 from /usr/include/openssl/bio.h:30,
                 from /usr/include/openssl/conf.h:24,
                 from tls.cpp:39:
/usr/include/openssl/types.h:134:16: note: class type ‘DH {aka dh_st}’ is incomplete
 typedef struct dh_st DH;
                ^~~~~
tls.cpp:363:20: error: cannot convert ‘DH* {aka dh_st*}’ to ‘BIGNUM* {aka bignum_st*}’ for argument ‘1’ to ‘void BN_free(BIGNUM*)’
          BN_free(dh);
                    ^
In file included from /usr/include/openssl/crypto.h:36:0,
                 from /usr/include/openssl/bio.h:30,
                 from /usr/include/openssl/conf.h:24,
                 from tls.cpp:39:
/usr/include/openssl/types.h:134:16: note: class type ‘DH {aka dh_st}’ is incomplete
 typedef struct dh_st DH;
                ^~~~~
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-atomic-alignment’
cc1plus: warning: unrecognized command line option ‘-Wno-address-of-packed-member’
make[1]: *** [Makefile:451: tls.o] Error 1
make[1]: Leaving directory '/root/KeyDB/src'
make: *** [Makefile:9: install] Error 2
#### 

OPENSSL_CFLAGS=-I/usr/include/openssl OPENSSL_LDFLAGS=-L/lib64 LDFLAGS=-L/lib64 LDLIBS=-lssl -lcrypto make USE_SYSTEMD=yes BUILD_TLS=yes install

**Comment by dirkesquire** [comment]
I am getting the same problem. I'm currently unable to build on Raspberry Pi 3 using Debian stretch.

The steps I followed was:
```sudo apt-get update
sudo apt-get install build-essential nasm autotools-dev autoconf libjemalloc-dev tcl tcl-dev uuid-dev libcurl4-openssl-dev git
git clone https://github.com/EQ-Alpha/KeyDB.git
cd KeyDB
make
sudo make install```

And the error is:

```pi@Pi3:~/Builds/KeyDB $ make
cd src && make all
make[1]: Entering directory '/home/pi/Builds/KeyDB/src'
    CC Makefile.dep
    CC object.o
object.cpp: In function ‘char* serializeStoredStringObject(sds, robj_roptr)’:
object.cpp:1585:5: error: static assertion failed
     static_assert((sizeof(robj) + sizeof(mvcc)) == sizeof(redisObjectStack), "");
     ^~~~~~~~~~~~~
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-atomic-alignment’
cc1plus: warning: unrecognized command line option ‘-Wno-address-of-packed-member’
Makefile:482: recipe for target 'object.o' failed
make[1]: *** [object.o] Error 1
make[1]: Leaving directory '/home/pi/Builds/KeyDB/src'
Makefile:6: recipe for target 'all' failed
make: *** [all] Error 2
pi@Pi3:~/Builds/KeyDB $ openssl version -a
OpenSSL 1.1.0l  10 Sep 2019
built on: reproducible build, date unspecified
platform: debian-armhf
options:  bn(64,32) rc4(char) des(long) blowfish(ptr)
compiler: gcc -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/lib/ssl\"" -DENGINESDIR="\"/usr/lib/arm-linux-gnueabihf/engines-1.1\""
OPENSSLDIR: "/usr/lib/ssl"
ENGINESDIR: "/usr/lib/arm-linux-gnueabihf/engines-1.1"```

The interesting thing is it looks like i'm getting the same problem with OpenSSL 1.1 if I'm not mistaken. Assuming that the build uses the system Openssl.

**Comment by dirkesquire** [comment]
Looking at the code in object.cpp:

```
sds serializeStoredStringObject(sds str, robj_roptr o)
{
    uint64_t mvcc;
    mvcc = mvccFromObj(o);
    str = sdscatlen(str, &mvcc, sizeof(mvcc));
    str = sdscatlen(str, &(*o), sizeof(robj));
    static_assert((sizeof(robj) + sizeof(mvcc)) == sizeof(redisObjectStack), "");
    switch (o->encoding)
    {
    case OBJ_ENCODING_EMBSTR:
    case OBJ_ENCODING_RAW:
        str = sdscatsds(str, (sds)szFromObj(o));
        break;

    case OBJ_ENCODING_INT:
        break;  //nop
    }
        
    return str;
}
```

Not sure this has anything to do with OpenSSL at all in fact.

**Comment by dirkesquire** [comment]
The only reference to "-Wno-atomic-alignment" I can find in the entire repository is in the makefile:

```
# Default settings
STD=-pedantic -DREDIS_STATIC=''
CXX_STD=-std=c++14 -pedantic -fno-rtti -D__STDC_FORMAT_MACROS  
ifneq (,$(findstring clang,$(CC)))
  STD+=-Wno-c11-extensions
else
ifneq (,$(findstring FreeBSD,$(uname_S)))
  STD+=-Wno-c11-extensions
endif
endif
WARN=-Wall -W -Wno-missing-field-initializers -Wno-address-of-packed-member -Wno-atomic-alignment
OPT=$(OPTIMIZATION)
```

**Comment by msotheeswaran** [comment]
Fixed in 6.3.2!
